### PR TITLE
Datasource for transit gateway

### PR DESCRIPTION
**Pull request by pkrishn1-pk**

### CODE REVIEW

File: ibm/data_source_ibm_tg_gateways.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.ibm.com/ibmcloud/networking-go-sdk/transitgatewayapisv1"
+)
+
+func dataSourceIBMTransitGateways() *schema.Resource {
+	log.Println("inside of transit gateways:")
+	return &schema.Resource{
+		Read: dataSourceIBMTransitGatewaysRead,
+		Schema: map[string]*schema.Schema{
+			"id": {
```
Comment (hkantare) [inline ibm/data_source_ibm_tg_gateways.go:16]: Remove this "id" ...don't declate terraform will take care of id d.SetId(..)

File: ibm/data_source_ibm_tg_gateways.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+func dataSourceIBMTransitGatewaysRead(d *schema.ResourceData, meta interface{}) error {
+	log.Println("first line in read")
+	client, err := transitgatewayClient(meta)
+	if err != nil {
+		return err
+	}
+
+	log.Println("passed getting transitgateway")
+	listTransitGatewaysOptionsModel := &transitgatewayapisv1.ListTransitGatewaysOptions{}
+	//listTransitGatewaysOptionsModel.Headers = map[string]string{"x-custom-header": "x-custom-value"}
```
Comment (hkantare) [inline ibm/data_source_ibm_tg_gateways.go:78]: Remove the unnecessary log stmts

File: website/docs/d/tg_gateway.html.markdown [hunk 3]
```diff
@@ -0,0 +1,48 @@
+---
+layout: "ibm"
+page_title: "IBM : tg_gateways"
```
Comment (hkantare) [inline website/docs/d/tg_gateway.html.markdown:3]: it should be tg_gateway

File: website/docs/d/tg_gateway.html.markdown [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+The following arguments are supported:
+
+* `id` - (Required, string) The id of the gateway.
+
+## Attribute Reference
+
+The following attributes are exported:
+
+* `gateways` - List of all Transit gateways in the IBM Cloud Infrastructure.
```
Comment (hkantare) [inline website/docs/d/tg_gateway.html.markdown:40]: The attributes are not part of gateways(list).All are higher level attributes

File: website/docs/d/tg_gateways.html.markdown [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+layout: "ibm"
+page_title: "IBM : tg_gateways"
+sidebar_current: "docs-ibm-datasource-tg-gateways"
+description: |-
+  Manages IBM Cloud Infrastructure Transit Gateway.
+---
+
+# ibm\_tg_gateways
+
+Import the details of an existing IBM Cloud Infrastructure transit gateway as a read-only data source. You can then reference the fields of the data source in other resources within the same configuration using interpolation syntax.
```
Comment (hkantare) [inline website/docs/d/tg_gateways.html.markdown:11]: change gateway to gateways

File: ibm/data_source_ibm_tg_gateway.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func dataSourceIBMTransitGateway() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceIBMTransitGatewayRead,
+
+		Schema: map[string]*schema.Schema{
+
+			tgID: {
```
Comment (hkantare) [inline ibm/data_source_ibm_tg_gateway.go:15]: Generally we write datasources to take the name of the resource ..
Try to fetch all gateways and match with name

File: ibm/data_source_ibm_tg_gateways.go (hunk 1 above)
Comment (pkrishn1-pk) [reply ibm/data_source_ibm_tg_gateways.go]: removed.

File: website/docs/d/tg_gateway.html.markdown (hunk 3 above)
Comment (pkrishn1-pk) [reply website/docs/d/tg_gateway.html.markdown]: changed

File: website/docs/d/tg_gateways.html.markdown (hunk 5 above)
Comment (pkrishn1-pk) [reply website/docs/d/tg_gateways.html.markdown]: changed.

File: ibm/data_source_ibm_tg_gateways.go (hunk 2 above)
Comment (pkrishn1-pk) [reply ibm/data_source_ibm_tg_gateways.go]: removed log statements.

File: website/docs/d/tg_gateway.html.markdown (hunk 4 above)
Comment (pkrishn1-pk) [reply website/docs/d/tg_gateway.html.markdown]: yeah..removed it.

File: ibm/data_source_ibm_tg_gateway.go (hunk 6 above)
Comment (pkrishn1-pk) [reply ibm/data_source_ibm_tg_gateway.go]: ok.. changed to take name instead of id.

File: ibm/data_source_ibm_tg_gateways_test.go [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+					resource.TestCheckResourceAttrSet(resName, "transit_gateways.0.location"),
+					resource.TestCheckResourceAttrSet(resName, "transit_gateways.0.global"),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckIBMTransitGatewaysDataSourceConfig() string {
+	// status filter defaults to empty
```
Comment (hkantare) [inline ibm/data_source_ibm_tg_gateways_test.go:30]: Does this test case works if you don't have any gateways in our account I think  it fails at this condition checks
resource.TestCheckResourceAttrSet(resName, "transit_gateways.0.name"),

Change the testcase to create and then fetch the gateways some reference
https://github.com/IBM-Cloud/terraform-provider-ibm/blob/master/ibm/data_source_ibm_is_subnets_test.go

File: ibm/data_source_ibm_tg_gateway.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+			d.Set("location", tgw.Location)
+			d.Set("created_at", tgw.CreatedAt.String())
+
+			if tgw.UpdatedAt != nil {
+				d.Set("updated_at", tgw.UpdatedAt.String())
+			}
+			d.Set("global", tgw.Global)
+			d.Set("status", tgw.Status)
+		}
+	}
```
Comment (hkantare) [inline ibm/data_source_ibm_tg_gateway.go:90]: Add  a variable found bool with default false....and check if found false then retrun err could n't find any gateway with specified name

File: ibm/data_source_ibm_tg_gateway.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+func dataSourceIBMTransitGateway() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceIBMTransitGatewayRead,
+
+		Schema: map[string]*schema.Schema{
+			tgName: {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			tgID: {
```
[the remark this answers is not in the document]
Comment (pkrishn1-pk) [reply ibm/data_source_ibm_tg_gateway.go]: Removed.

File: ibm/data_source_ibm_tg_gateway.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceIBMTransitGatewayRead(d *schema.ResourceData, meta interface{}) error {
+
+	/*
+		id := d.Get(tgName).(string)
```
[the remark this answers is not in the document]
Comment (pkrishn1-pk) [reply ibm/data_source_ibm_tg_gateway.go]: Removed.

File: ibm/data_source_ibm_tg_gateways.go [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+func dataSourceIBMTransitGatewaysRead(d *schema.ResourceData, meta interface{}) error {
+
+	client, err := transitgatewayClient(meta)
+	if err != nil {
+		return err
+	}
+
+	listTransitGatewaysOptionsModel := &transitgatewayapisv1.ListTransitGatewaysOptions{}
+	listTransitGateways, _, err := client.ListTransitGateways(listTransitGatewaysOptionsModel)
```
[the remark this answers is not in the document]
Comment (pkrishn1-pk) [reply ibm/data_source_ibm_tg_gateways.go]: Added more details in error response.

File: ibm/data_source_ibm_tg_gateway.go [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+			return err
+		}
+	*/
+	client, err := transitgatewayClient(meta)
+	if err != nil {
+		return err
+	}
+
+	listTransitGatewaysOptionsModel := &transitgatewayapisv1.ListTransitGatewaysOptions{}
+	listTransitGateways, _, err := client.ListTransitGateways(listTransitGatewaysOptionsModel)
```
[the remark this answers is not in the document]
Comment (pkrishn1-pk) [reply ibm/data_source_ibm_tg_gateway.go]: Added more details in error response.

File: ibm/data_source_ibm_tg_gateways_test.go (hunk 7 above)
Comment (pkrishn1-pk) [reply ibm/data_source_ibm_tg_gateways_test.go:43]: Added code to create and check gateways list. Tested with no transit gateways in account and test case is working fine.

File: ibm/data_source_ibm_tg_gateway.go (hunk 8 above)
Comment (pkrishn1-pk) [reply ibm/data_source_ibm_tg_gateway.go:87]: added a check for gateway found or not and then reply error message if not found.

File: ibm/data_source_ibm_tg_gateways.go [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+import (
+	"log"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.ibm.com/ibmcloud/networking-go-sdk/transitgatewayapisv1"
+)
+
+func dataSourceIBMTransitGateways() *schema.Resource {
+	log.Println("inside of transit gateways:")
```
[the remark this answers is not in the document]
Comment (pkrishn1-pk) [reply ibm/data_source_ibm_tg_gateways.go]: Removed

**Pull request merged**